### PR TITLE
SSL w/o torchaudio dependency

### PR DESCRIPTION
**Pull request by wanchichen**
## What?

This PR allows for HuBERT pre-training w/o using torchaudio, allowing for more customization and use of different ESPnet components. It also introduces some tricks to better support large-scale training. 

Features:
- Flash Attention (by @pyf98)
- Activation checkpointing in E-Branchformer
- Convolutional Feature Extractor as frontend
- Convolutional positional embeddings
- Efficient batch sampler for large-scale
- Efficient multi-shard iterator for large-scale
- WavLM noise augmentation
- More efficient distributed training
- Recommended tunable DDP args for multi-node training
- Cross-Entropy loss for HuBERT SSL
- Intermediate losses for HuBERT SSL
- HuBERT SSL using filterbank input features
- More detailed GPU memory reporting

Supports only HuBERT for Transformer and E-Branchformer so far. 

To do:

- [ ] Fbank training configs
- [ ] Conformer, branchformer encoder
- [ ] check in `hubert.sh` which implementation to use

### CODE REVIEW

**Comment by sw005320** [comment]
@wanchichen, can you restart this PR?

**Comment by sw005320** [comment]
@wanchichen, let’s finish this PR.
There are a lot of conflicts now. So, please resolve them.

@simpleoier, please also review this PR.

File: espnet2/train/trainer.py [hunk 1]
```diff
@@ -512,11 +520,6 @@ def train_one_epoch(
         ):
             assert isinstance(batch, dict), type(batch)
 
-            if distributed:
```
Comment (wanchichen) [inline espnet2/train/trainer.py:515]: I performed several experiments (ASR, SSL) and found that this was useless. But we may need to experiment with other batch samplers to make sure

File: espnet2/train/trainer.py [hunk 2]
```diff
@@ -709,7 +731,7 @@ def train_one_epoch(
                 for iopt, optimizer in enumerate(optimizers):
                     if optim_idx is not None and iopt != optim_idx:
                         continue
-                    optimizer.zero_grad()
+                    optimizer.zero_grad(set_to_none=True)
```
Comment (wanchichen) [inline espnet2/train/trainer.py:734]: This is the recommended setting by PyTorch now, it is both faster and more memory efficient. However, it may also slightly affect the learning curve

File: espnet/nets/pytorch_backend/transformer/attention.py [hunk 3]
```diff
@@ -106,6 +126,61 @@ def forward(self, query, key, value, mask):
...
+                    )   # (batch_size, seqlen, nheads, headdim)
+                    out = out.reshape(out.shape[0], out.shape[1], -1)
+                    out = self.linear_out(out)
+                    return out
+            except Exception as e:
+                if self.training:
+                    import logging; logging.warning(f"Flash attn has exception: {e}")
+                pass
+
+        # import logging; logging.warning("ESPnet attention")
```
Comment (simpleoier) [inline espnet/nets/pytorch_backend/transformer/attention.py:182]: clean this?

File: espnet2/train/trainer.py [hunk 4]
```diff
@@ -241,7 +242,11 @@ def run(
                         else None
                     ),
                     find_unused_parameters=trainer_options.unused_parameters,
+                    static_graph=True,
```
Comment (wanchichen) [inline espnet2/train/trainer.py:245]: I think `gradient_as_bucket_view` can be left as default, but `static_graph` and ` dp_model.register_comm_hook(None, bf16_compress_hook)` may need to be configurable flags by the user. Any suggestions?

File: espnet/nets/pytorch_backend/transformer/embedding.py [hunk 5]
```diff
@@ -383,3 +383,60 @@ def forward(self, x: torch.Tensor, start_idx: int = 0):
         self.extend_pe(x.size(1) + start_idx, x.device, x.dtype)
         x = x * self.xscale + self.pe[:, start_idx : start_idx + x.size(1)]
         return self.dropout(x)
+
+class ConvolutionalPositionalEmbedding(torch.nn.Module):
+    """Positional embedding which is placed at the beginning of Transformer.
```
Comment (simpleoier) [inline espnet/nets/pytorch_backend/transformer/embedding.py:388]: Make it more specific explanation about Convolutional PosEmb?

File: espnet2/asr/encoder/e_branchformer_encoder.py [hunk 6]
```diff
@@ -154,6 +166,7 @@ def forward(self, x_input, mask, cache=None):
 
         if pos_emb is not None:
             x2 = (x2, pos_emb)
+        #x2 = torch.utils.checkpoint.checkpoint(self.cgmlp, x2, mask, use_reentrant=False)
```
Comment (simpleoier) [inline espnet2/asr/encoder/e_branchformer_encoder.py:169]: clean this?

File: espnet2/asr/frontend/cnn.py
Comment (simpleoier) [inline espnet2/asr/frontend/cnn.py:1]: I suggest to make the name easy to understand.

File: espnet2/asr/preencoder/subsampling.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+from typing import Tuple
+
+import torch
+from typeguard import check_argument_types
+
+from espnet2.asr.preencoder.abs_preencoder import AbsPreEncoder
+from espnet.nets.pytorch_backend.transformer.subsampling import Conv2dSubsampling
+
+
+class Subsampling(AbsPreEncoder):
```
Comment (simpleoier) [inline espnet2/asr/preencoder/subsampling.py:16]: more precise, e.g. Conv2dSubsampling?

File: espnet2/bin/launch.py
Comment (simpleoier) [inline espnet2/bin/launch.py:1]: What change does this file include?

File: espnet2/bin/ssl_train.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+def get_parser():
+    parser = SSLTask.get_parser()
+    return parser
+
+
+def main(cmd=None):
+    """Hubert pretraining.
+
+    Example:
+        % python hubert_train.py asr --print_config --optim adadelta \
```
Comment (simpleoier) [inline espnet2/bin/ssl_train.py:14]: Is this a discrepancy between the example and the filename?

File: espnet2/fileio/read_text.py [hunk 9]
```diff
@@ -32,7 +32,7 @@ def read_2columns_text(path: Union[Path, str]) -> Dict[str, str]:
                 k, v = sps
 
             if k in data:
-                raise RuntimeError(f"{k} is duplicated ({path}:{linenum})")
+                logging.warning(f"{k} is duplicated ({path}:{linenum})")
```
Comment (simpleoier) [inline espnet2/fileio/read_text.py:35]: I am not confident about this change. @sw005320 Can you double-check this?

File: espnet2/s2t/espnet_model.py
Comment (simpleoier) [inline espnet2/s2t/espnet_model.py:1]: Is this file related to the purpose of this PR?

File: espnet2/ssl/loss/hubert_loss_ce.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                loss_u = F.cross_entropy(hs_u, targets_u.long(), reduction='sum') * self.unmasked_loss_weight * self.layer_weights[i]
+                losses_u.append(loss_u.detach().item())
+                total_loss += loss_u
+                
+        total_loss = feature_penalty * feature_weight * hs_m.shape[0] + total_loss
+
+        correct_m, count_m = self._compute_correct(hs_m, targets_m)
+        correct_u, count_u = self._compute_correct(hs_u, targets_u)
+
+        #    hubert_losses_m=losses_m[0] if len(losses_m) > 0 else None,
```
Comment (simpleoier) [inline espnet2/ssl/loss/hubert_loss_ce.py:142]: clean this?

File: espnet2/tasks/abs_task.py [hunk 11]
```diff
@@ -1880,7 +1885,7 @@ def build_multiple_iter_factory(
             for i in range(num_splits)
         ]
         num_iters_per_epoch_list = [
-            (iter_options.num_iters_per_epoch + i) // num_splits
+            (iter_options.num_iters_per_epoch + i) # // num_splits
```
Comment (simpleoier) [inline espnet2/tasks/abs_task.py:1888]: Is this change an intended one or for the debugging purpose?

File: espnet2/train/preprocessor.py [hunk 12]
```diff
@@ -392,6 +392,7 @@ def _speech_process(
             if self.speech_volume_normalize is not None:
                 speech = data[self.speech_name]
                 ma = np.max(np.abs(speech))
+                print(speech.shape, self.speech_volume_normalize, ma, flush=True)
```
Comment (simpleoier) [inline espnet2/train/preprocessor.py:395]: clean this?

File: espnet2/train/trainer.py [hunk 13]
```diff
@@ -627,7 +631,25 @@ def train_one_epoch(
                     # Backward passes under autocast are not recommended.
                     # Backward ops run in the same dtype autocast chose
                     # for corresponding forward ops.
-                    scaler.scale(loss).backward()
+                    try:
+                        scaler.scale(loss).backward()
+                    except:
+                        torch.cuda.synchronize()
+                        print("OOM in backwards. Retrying after emptying cache.", flush=True)
```
Comment (simpleoier) [inline espnet2/train/trainer.py:636]: Use logging?

File: espnet2/tasks/abs_task.py (hunk 11 above)
Comment (wanchichen) [reply espnet2/tasks/abs_task.py]: Its intended, just added a flag that details it

**Comment by sw005320** [comment]
Can you also fix the conflicts?
Then, CI will start.

File: espnet2/s2t/espnet_model.py
Comment (wanchichen) [reply espnet2/s2t/espnet_model.py:1]: Nope, this is an outdated API change

File: espnet2/fileio/read_text.py (hunk 9 above)
Comment (wanchichen) [reply espnet2/fileio/read_text.py:35]: We could keep it as an error, although the warning is useful in large multi-corpora work if you accidentally have some overlap and filtering manually is too expensive.

**Comment by wanchichen** [comment]
> Thanks! I made some comments. One question I have is about the `espnet2/ssl/mask`. What is the benefit of a new mask module? If it is only used in HuBERT models during pre-training, a single function in the hubert model would be enough.

Thanks! I was trying to consider the case where different models have different masking/training strategies, for future-proofing. For example, data2vec 2.0 removes the masked tokens, or a wav2vec2 masker may need to return the positive/negative indices. 

We could do all of this within a single module, but I figured breaking it up would be easier to read/navigate. Do you have any recommendations?

**Comment by sw005320** [comment]
Can you fix the CI errors?

**Comment by simpleoier** [comment]
> We could do all of this within a single module, but I figured breaking it up would be easier to read/navigate. Do you have any recommendations?

I got it. A single mask module might be okay.

File: egs2/ml_superb/asr1/local/visualizer.ipynb
Comment (sw005320) [inline egs2/ml_superb/asr1/local/visualizer.ipynb:1]: Do you need this?

File: espnet/nets/pytorch_backend/nets_utils.py
Comment (sw005320) [inline espnet/nets/pytorch_backend/nets_utils.py:1]: Can you make a test for this?
Is it difficult to make it?

This function is very buggy, and I want to ensure your change is safe.

**Comment by sw005320** [comment]
@wanchichen,
- can you try to finish this as soon as possible?
- can you separate the flash attention implementation with another PR?

**Comment by wanchichen** [comment]
Yeah I am gonna break up this PR, its far too large now.

I will update this comment as i open individual ones.

- [ ] Flash Attention
- [ ] SSL architecture
- [ ] SSL Recipe
- [ ] Trainer
- [ ] Batching

File: espnet2/asr/encoder/e_branchformer_encoder.py [hunk 14]
```diff
@@ -468,11 +522,13 @@ def forward(
...
-                        break
+            for layer_idx, encoder_layer in enumerate(self.encoders):
+                xs_pad, masks = encoder_layer(xs_pad, masks)
+                if return_all_hs:
+                    if isinstance(xs_pad, tuple):
+                        intermediate_outs.append(xs_pad[0])
+                    else:
+                        intermediate_outs.append(xs_pad)
             else:
                 xs_pad, masks = self.encoders(xs_pad, masks)
```
Comment (seastar105) [inline espnet2/asr/encoder/e_branchformer_encoder.py:533]: @wanchichen 
I found this for-else makes encoders forward pass doubled while trying XEUS model. Is it intended or bug?

Comment (wanchichen) [reply espnet2/asr/encoder/e_branchformer_encoder.py:533]: Oh this is a bug. Good catch! I will need to check how this impacts performance

**Comment by rumourscape** [comment]
Is this going to be merged?

**Comment by sw005320** [comment]
@wanchichen, can you revisit this PR?
Now, flash attention is merged.
You may restart this PR or make a new PR.

File: espnet2/asr/encoder/e_branchformer_encoder.py (hunk 14 above)
Comment (seastar105) [reply espnet2/asr/encoder/e_branchformer_encoder.py:533]: @wanchichen any update on this? i wanna try XEUS as feature extractor
